cdc/owner.go

```diff
@@ -718,13 +739,30 @@ func (o *Owner) flushChangeFeedInfos(ctx context.Context) error {
 		return nil
 	}
 
-	minCheckpointTs := uint64(math.MaxUint64)
+	staleChangeFeeds := make(map[model.ChangeFeedID]*model.ChangeFeedStatus, len(o.changeFeeds))
+	gcSafePoint := uint64(math.MaxUint64)
+
+	// get the lower bound of gcSafePoint
+	minGCSafePoint, err := o.getGcMinSafePointCache(ctx)
+	if err != nil {
+		log.Warn("failed to acquire minGCSafePoint from Cache, will use this machine time", zap.Error(err))
+		minGCSafePoint = oracle.EncodeTSO(oracle.GetPhysical(time.Now().Add(-(time.Second * time.Duration(o.gcTTL)))))
+	}
+
 	if len(o.changeFeeds) > 0 {
 		snapshot := make(map[model.ChangeFeedID]*model.ChangeFeedStatus, len(o.changeFeeds))
 		for id, changefeed := range o.changeFeeds {
 			snapshot[id] = changefeed.status
-			if changefeed.appliedCheckpointTs < minCheckpointTs {
-				minCheckpointTs = changefeed.appliedCheckpointTs
+			if changefeed.status.CheckpointTs < gcSafePoint {
+				gcSafePoint = changefeed.status.CheckpointTs
+			}
+			// If changefeed's appliedCheckpoinTs < minGCSafePoint, it means this changefeed is stagnant.
+			// They are collected into this map, and then handleStaleChangeFeed() is called to deal with these stagnant changefeed.
+			// A changefeed will not enter the map twice, because in run(),
+			// handleAdminJob() will always be executed before flushChangeFeedInfos(),
+			// ensuring that the previous changefeed in staleChangeFeeds has been stopped and removed from o.changeFeeds.
+			if changefeed.status.CheckpointTs < minGCSafePoint {
+				staleChangeFeeds[id] = changefeed.status
 			}
```

Comment:
If the `CheckpointTs` of changefeed is less than `minGCSafePoint`, `gcSafePoint` is still set to a value less than `minGCSafePoint`?

Reply:
Yes. If this happens, in order to prevent the running `changefeed` data loss , the `gcSafePoint` will be set to `CheckpointTs` in first. 
In the same time,  the `changefeed` will be put into `staleChangeFeeds`.  
The update time of `gcSafePoint` will actually be delayed until the next `flushChangeFeedInfos()` excution.
In the next time `flushChangeFeedInfos()` is excuted the `changefeed` is stopped, the value of `gcSafePoint` will really push forward.
For summary, `minGCSafePoints` just a lower bound for judging whether a `changefeed` is stagnant, and the value of `gcSafePoint` does not directly depend on it.